factorio: update all versions

ikervagyok opened the pull request:
also adds libpulseaudio, as experimental already needs is and soon 
stable will need it too.

###### Motivation for this change
update

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Baughn @elitak

ikervagyok committed:
factorio: update all versions to  0.17.79

and adds libpulseaudio as a dependency

Baughn commented on pkgs/games/factorio/default.nix:
```diff
@@ -175,10 +175,11 @@ let
 
       libPath = stdenv.lib.makeLibraryPath [
         alsaLib
+        libpulseaudio
```
This addition probably means 17.x won't also need alsaLib. Can you check that, and if it's true, add a TODO to remove it when stable is updated?

elitak commented:
Looks good; built and ran fine for me.

filalex77 commented on pkgs/games/factorio/default.nix:
```diff
@@ -52,15 +52,15 @@ let
       alpha = {
-        stable        = bdist { sha256 = "0b4hbpdcrh5hgip9q5dkmw22p66lcdhnr0kmb0w5dw6yi7fnxxh0"; version = "0.16.51"; withAuth = true; };
-        experimental  = bdist { sha256 = "1q66chnxsdlaz1bj3al62iikyxvknj1vkwh5bcc46favy4wpqpzz"; version = "0.17.52"; withAuth = true; };
+        stable        = bdist { sha256 = "0qasp5qm3q6zg4l077dpyyg7gx3mm0zszhvi41g95611hzrksrh6"; version = "0.17.74"; withAuth = true; };
```
```suggestion
        stable        = bdist { sha256 = "1p5zjib46d76ngyfwn593a3mhqn71nz9kyyxfv87z8xkhj5a47jf"; version = "0.17.74"; withAuth = true; };
```

filalex77 commented on pkgs/games/factorio/default.nix:
```diff
@@ -52,15 +52,15 @@ let
-        stable        = bdist { sha256 = "0b4hbpdcrh5hgip9q5dkmw22p66lcdhnr0kmb0w5dw6yi7fnxxh0"; version = "0.16.51"; withAuth = true; };
-        experimental  = bdist { sha256 = "1q66chnxsdlaz1bj3al62iikyxvknj1vkwh5bcc46favy4wpqpzz"; version = "0.17.52"; withAuth = true; };
+        stable        = bdist { sha256 = "0qasp5qm3q6zg4l077dpyyg7gx3mm0zszhvi41g95611hzrksrh6"; version = "0.17.74"; withAuth = true; };
+        experimental  = bdist { sha256 = "13ck7sabnl3k8i4cr4hxzschkvlikwzcpd04rf0iv9j89q08dzly"; version = "0.17.76"; withAuth = true; };
```
```suggestion
        experimental  = bdist { sha256 = "1qx86n1y5wc769dbr4ciia08zq45c8dss3i6qzvppnz18fs4b4vr"; version = "0.17.76"; withAuth = true; };
```

Br1ght0ne reviewed (changes requested):
Got two hash mismatches:

```
hash mismatch in fixed-output derivation '/nix/store/d8kwrnybxd2cnml708xrags1kgdv7z0m-factorio_alpha_x64-0.17.74.tar.xz':
  wanted: sha256:0qasp5qm3q6zg4l077dpyyg7gx3mm0zszhvi41g95611hzrksrh6
  got:    sha256:1p5zjib46d76ngyfwn593a3mhqn71nz9kyyxfv87z8xkhj5a47jf
hash mismatch in fixed-output derivation '/nix/store/hjr3h3ii5rk0wmxgqgd72433211b4rhz-factorio_alpha_x64-0.17.76.tar.xz':
  wanted: sha256:13ck7sabnl3k8i4cr4hxzschkvlikwzcpd04rf0iv9j89q08dzly
  got:    sha256:1qx86n1y5wc769dbr4ciia08zq45c8dss3i6qzvppnz18fs4b4vr
```

Other than that, builds and runs fine.

ikervagyok commented:
@filalex77: you're right, i've fixed the hashes and updated to the latest release. also built on my own hydra now, to check automatically if i messed up.

@Baughn: i've tested this and we need alsa as well as pulse.

filalex77 commented:
@GrahamcOfBorg build factorio factorio-demo factorio-experimental factorio-headless factorio-headless-experimental

elitak commented:
Looks okay. Do I need to tell the bot somehow?

filalex77 commented:
@elitak These checks are not *required* to merge - they're just a helpful reference; thus, if you tested it, one of the maintainers should merge it when possible.

TheSirC commented:
I have a weird error popping up after installing the derivation :

# Description of the error
After building & installing the game and trying to start a new campaign, I get the following popup in game and in the console : 

	106.474 Error MainLoop.cpp:1195: Exception at tick 5158719: Error while running reload_script: 
	filesystem error: copy_file "/nix/store/dmsmbfvhsfgybbldgsmfczy2mj8y9jkg-factorio-demo-0.17.79/share/factorio/data/base/campaigns/npe/locale/uk/npe.cfg" to "/home/sirc/.factorio/temp/currently-playing/campaign-locale/uk/npe.cfg" failed: 
	Permission denied [/nix/store/dmsmbfvhsfgybbldgsmfczy2mj8y9jkg-factorio-demo-0.17.79/share/factorio/data/base/campaigns/npe/locale/uk/npe.cfg] [/home/sirc/.factorio/temp/currently-playing/campaign-locale/uk/npe.cfg]

# System
From the binary itself : 

	   0.000 2019-12-23 22:20:12; Factorio 0.17.79 (build 47865, linux64, demo)

# Steps to reproduce

1. Download the nix file
2. `nix repl`
3. `pkgs = import <nixpkgs> {}`
1. `factorio = pkgs.callPackages factorio.nix {}`
1. `:i factorio`
1. `:q `
1. `factorio` (as non-root)
1. Try to launch a new campaign

# Notes
Of course, the error seems to stem from filesystem permission issues but I launch it as a normal user (the person that owns the home directory) so _is this the expected behavior ?_

Of course, launching the game as root solves the issue as expected but _should you have to launch the game as root ?_

worldofpeace commented:
> Of course, launching the game as root solves the issue as expected but should you have to launch the game as root ?

Absolutely not, and I'm just assuming that because why would a game ever need to have such privilege?

> Permission denied [/nix/store/dmsmbfvhsfgybbldgsmfczy2mj8y9jkg-factorio-demo-0.17.79/share/factorio/data/base/campaigns/npe/locale/uk/npe.cfg] [/home/sirc/.factorio/temp/currently-playing/campaign-locale/uk/npe.cfg]

This is actually a common problem that happens with NixOS.
It sounds like factorio copies these files from the nix store to some temp directory and wants to write to them, but it doesn't chmod them and make them owned by the proper user. I bet you can fix this on your own by fixing the permissions, and this is a pretty reasonable bug report to submit (I make one for it all the time).

worldofpeace commented:
This has got some testing in https://github.com/NixOS/nixpkgs/pull/77288, so I don't think there's much issues. Will merge.

TheSirC commented:
>  bet you can fix this on your own by fixing the permissions, and this is a pretty reasonable bug report to submit (I make one for it all the time).

I will look into how the game developers want bug reports then.

> I bet you can fix this on your own by fixing the permissions.

I will try then. I have just no idea which user should be given the permissions.

worldofpeace commented:
@TheSirC You probably want to chown /home/sirc/.factorio/ to your user and group, and make sure all the files have the correct writ-ability.

TheSirC commented:
> You probably want to chown /home/sirc/.factorio/ to your user and group

@worldofpeace That is what is weird to me because `ls -laR .factorio` sends back

<details>
<summary>Output of the command</summary>

```
.factorio:
total 788
drwxr-xr-x  4 sirc users   4096 23 déc.  22:39 .
drwx------ 46 sirc users   4096 27 déc.  10:12 ..
-rw-r--r--  1 sirc users     19 23 déc.  22:39 achievements.dat
-rw-r--r--  1 sirc users   8777 15 déc.  15:56 blueprint-storage-demo.dat
-rwxr-xr-x  1 sirc users  23038 23 déc.  22:30 config.cfg
-rw-r--r--  1 sirc users 727369 23 déc.  22:30 crop-cache.dat
-rw-r--r--  1 sirc users   4144 23 déc.  22:39 factorio-current.log
-rw-r--r--  1 sirc users   4148 23 déc.  22:27 factorio-previous.log
drwxr-xr-x  2 sirc users   4096 23 déc.  22:39 mods
-rw-r--r--  1 sirc users    959 23 déc.  22:39 player-data.json
drwxr-xr-x  2 sirc users   4096 15 déc.  15:55 saves

.factorio/mods:
total 12
drwxr-xr-x 2 sirc users 4096 23 déc.  22:39 .
drwxr-xr-x 4 sirc users 4096 23 déc.  22:39 ..
-rw-r--r-- 1 sirc users   85 23 déc.  22:39 mod-list.json

.factorio/saves:
total 8
drwxr-xr-x 2 sirc users 4096 15 déc.  15:55 .
drwxr-xr-x 4 sirc users 4096 23 déc.  22:39 ..
```

</details>


which states that all the files are already in my group's and in my user's permissions.

worldofpeace commented:
@TheSirC The directory that the error pointed doesn't exist according to the output of that command `temp/currently-playing/campaign-locale/uk/npe.cfg`

akamaus commented:
@TheSirC @worldofpeace I recently stumbled upon a same error after upgrading factorio to 0.18.18. I did some quick experiments and found that game removes files in `.factorio/temp/currently_playing` after you hit the button on error message window (which is logical, because you return to main menu and dir is called `currently_playing`). If you check the directory just after you get the error, you would see than npe.cfg file. it will be in `RR-` mode.

As a terrible workaround I just remounted `/nix/store` and manually did `chmod +rw` for `share` subfolder.

grahamc commented:
Remounting the nix store rw compromises the nix store and should not be done.

TheSirC commented:
> Remounting the nix store rw compromises the nix store and should not be done.

It did seem quite...sketchy (?)

What would be the recommended way to circumvent the issue then ?

akamaus commented:
@TheSirC. I see three possible ways. Only the third one being correct, imho.
* Somehow fix the permissions inside /nix/store. I guess it may be hard because AFAIK they're reset to 444 on last installation stage
* Implement some weird hook using LD_PRELOAD or something like that. We should detect copy being created and fix permissions on the fly.
* Solve the problem upstream, I would say it's perfectly ok to have master-copy of files read-only. If not, why to copy them in the first place? It's responsibility of Factorio to ensure paths it's going to write to have appropriate permissions.

elitak commented:
If you must write to the /nix/store path, you could have the wrapper script place either a bind mount or overlayfs over the directories that need write permission. Otherwise, I agree that getting the devs upstream to fix the issue is the best course. They are pretty responsive once you reach them. A post on their forums usually gets their attention.
